driver-core/src/main/com/mongodb/internal/session/ServerSessionPool.java

```diff
@@ -77,57 +72,46 @@ public ServerSessionPool(final Cluster cluster, @Nullable final ServerApi server
 
     public ServerSession get() {
         isTrue("server session pool is open", !closed);
-        ServerSessionImpl serverSession = serverSessionPool.get();
-        while (shouldPrune(serverSession)) {
-            serverSessionPool.release(serverSession, true);
-            serverSession = serverSessionPool.get();
+        ServerSessionImpl serverSession = available.pollLast();
+        while (serverSession != null && shouldPrune(serverSession)) {
+            serverSession.close();
+            serverSession = available.pollLast();
+        }
+        if (serverSession == null) {
+            serverSession = new ServerSessionImpl();
         }
+        inUseCount.incrementAndGet();
         return serverSession;
     }
 
     public void release(final ServerSession serverSession) {
-        serverSessionPool.release((ServerSessionImpl) serverSession);
-        serverSessionPool.prune();
+        inUseCount.decrementAndGet();
+        ServerSessionImpl serverSessionImpl = (ServerSessionImpl) serverSession;
+        if (serverSessionImpl.isMarkedDirty()) {
+            serverSessionImpl.close();
+        } else {
+            available.addLast(serverSessionImpl);
+        }
     }
 
     public void close() {
-        try {
-            closing = true;
-            serverSessionPool.close();
-
-            List<BsonDocument> identifiers;
-            synchronized (this) {
-                identifiers = new ArrayList<>(closedSessionIdentifiers);
-                closedSessionIdentifiers.clear();
-            }
-            endClosedSessions(identifiers);
-        } finally {
+        if (!closed) {
             closed = true;
+            List<BsonDocument> identifiers = new ArrayList<>(available.size());
+            ServerSessionImpl nextSession = available.pollFirst();
+            while (nextSession != null) {
+                nextSession.close();
+                identifiers.add(nextSession.getIdentifier());
+                nextSession = available.pollFirst();
+            }
```

Comment:
It appears that this code may be replaced with
```java
List<BsonDocument> identifiers = new ArrayList<>(available.size());
available.forEach(session -> {
    session.close();
    identifiers.add(session.getIdentifier());
});
```
`forEach` has a potential to be more efficient, and also it is shorter :)